heartbeat/impl.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2016 Prometheus Team
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package heartbeat
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/prometheus/alertmanager/config"
+	"github.com/prometheus/common/log"
+)
+
+type integration interface {
+	Heartbeat
+	name() string
+}
+
+const contentTypeJSON = "application/json"
+
+func Build(confs []*config.Heartbeat) map[string]Fanout {
+	res := map[string]Fanout{}
+
+	for _, nc := range confs {
+		var (
+			hb  = Fanout{}
+			add = func(i int, on integration) { hb[fmt.Sprintf("%s/%d", on.name(), i)] = on }
+		)
+
+		for i, c := range nc.OpsGenieConfigs {
+			n := NewOpsGenie(c)
+			add(i, n)
+		}
+		res[nc.Name] = hb
+	}
+	return res
+
+}
+
+// OpsGenie represents a OpsGenie implementation of Heartbeat
+type OpsGenie struct {
+	conf   *config.HeartbeatOpsGenieConfig
+	ticker *time.Ticker
+	done   chan struct{}
+
+	log log.Logger
+}
+
+// Returns a new OpsGenie object
+func NewOpsGenie(conf *config.HeartbeatOpsGenieConfig) *OpsGenie {
+	return &OpsGenie{
+		conf:   conf,
+		done:   make(chan struct{}),
+		ticker: time.NewTicker(time.Duration(conf.Interval)),
```

Comment:
Indeed I didn't see this... I made the Tick() method because it is not possible AFAIK to expose attributes on an interface. I could create a HeartbeatProvider interface and a Heartbeat struct to implement this or expose the interval in the HeartbeatRunner on a per implementation basis (map[string]struct{done chan, time.Duration interval})